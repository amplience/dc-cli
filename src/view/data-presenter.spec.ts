import DataPresenter from './data-presenter';

const stdoutWriteSpy = jest.spyOn(process.stdout, 'write');

interface TestItem {
  foo: string;
  key: string;
}

describe('DataPresenter', (): void => {
  beforeAll(() => {
    stdoutWriteSpy.mockImplementation(() => true);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  describe('single item', function() {
    const singleItem: TestItem = {
      foo: 'bar',
      key: 'value '.repeat(20)
    };

    it('should render a single item vertical table', (): void => {
      new DataPresenter(singleItem).render();
      expect(stdoutWriteSpy.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should render a single item as json', (): void => {
      new DataPresenter(singleItem).render({ json: true });
      expect(stdoutWriteSpy.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should render a single item as vertical table using tableUserConfig', (): void => {
      new DataPresenter(singleItem).render({
        tableUserConfig: {
          columns: {
            1: {
              width: 100
            }
          }
        }
      });
      expect(stdoutWriteSpy.mock.calls[0][0]).toMatchSnapshot();
    });
  });

  describe('collection of items tests', function() {
    describe('collection of items', function() {
      const collectionOfItems: TestItem[] = [
        {
          foo: 'bar1',
          key: 'value1 '.repeat(20)
        },
        {
          foo: 'bar2',
          key: 'value2 '.repeat(20)
        }
      ];

      it('should render a collection of items in a horizontal table', (): void => {
        new DataPresenter(collectionOfItems).render();
        expect(stdoutWriteSpy.mock.calls[0][0]).toMatchSnapshot();
      });

      it('should render a collection of items as json', (): void => {
        new DataPresenter(collectionOfItems).render({ json: true });
        expect(stdoutWriteSpy.mock.calls[0][0]).toMatchSnapshot();
      });

      it('should render using the supplied map() for a collection of items in a horizontal table', (): void => {
        new DataPresenter(collectionOfItems).render({ itemMapFn: ({ foo }: TestItem): object => ({ foo }) });
        expect(stdoutWriteSpy.mock.calls[0][0]).toMatchSnapshot();
      });

      it('should render a collection of items horizontal table using with table config', (): void => {
        new DataPresenter(collectionOfItems).render({
          tableUserConfig: {
            columns: {
              1: {
                width: 100
              }
            }
          }
        });
        expect(stdoutWriteSpy.mock.calls[0][0]).toMatchSnapshot();
      });
    });

    describe('collection of 0 items', function() {
      it('should render a collection of items in a horizontal table', (): void => {
        new DataPresenter([]).render();
        expect(stdoutWriteSpy.mock.calls[0][0]).toMatchSnapshot();
      });

      it('should render a collection of items as json', (): void => {
        new DataPresenter([]).render({ json: true });
        expect(stdoutWriteSpy.mock.calls[0][0]).toMatchSnapshot();
      });

      it('should render using the supplied map() for a collection of items in a horizontal table', (): void => {
        new DataPresenter([]).render({ itemMapFn: ({ foo }: TestItem): object => ({ foo }) });
        expect(stdoutWriteSpy.mock.calls[0][0]).toMatchSnapshot();
      });

      it('should render a collection of items horizontal table using with table config', (): void => {
        new DataPresenter([]).render({
          tableUserConfig: {
            columns: {
              1: {
                width: 100
              }
            }
          }
        });
        expect(stdoutWriteSpy.mock.calls[0][0]).toMatchSnapshot();
      });
    });
  });
});
