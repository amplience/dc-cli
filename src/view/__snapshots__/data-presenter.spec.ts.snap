// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataPresenter collection of items tests collection of 0 items should render a collection of items as json 1`] = `"[]"`;

exports[`DataPresenter collection of items tests collection of 0 items should render a collection of items horizontal table using with table config 1`] = `
"0 items returned.
"
`;

exports[`DataPresenter collection of items tests collection of 0 items should render a collection of items in a horizontal table 1`] = `
"0 items returned.
"
`;

exports[`DataPresenter collection of items tests collection of 0 items should render using the supplied map() for a collection of items in a horizontal table 1`] = `
"0 items returned.
"
`;

exports[`DataPresenter collection of items tests collection of items should render a collection of items as json 1`] = `
"[
  {
    \\"foo\\": \\"bar1\\",
    \\"key\\": \\"value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 \\"
  },
  {
    \\"foo\\": \\"bar2\\",
    \\"key\\": \\"value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 \\"
  }
]"
`;

exports[`DataPresenter collection of items tests collection of items should render a collection of items horizontal table using with table config 1`] = `
"+------+------------------------------------------------------------------------------------------------------+
| [1mfoo[22m  | [1mkey[22m                                                                                                  |
|------|------------------------------------------------------------------------------------------------------|
| bar1 | value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 va |
|      | lue1 value1 value1 value1 value1 value1                                                              |
|------|------------------------------------------------------------------------------------------------------|
| bar2 | value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 va |
|      | lue2 value2 value2 value2 value2 value2                                                              |
+------+------------------------------------------------------------------------------------------------------+

"
`;

exports[`DataPresenter collection of items tests collection of items should render a collection of items in a horizontal table 1`] = `
"+------+----------------------------------------------------------------------------------------------------------------------------------------------+
| [1mfoo[22m  | [1mkey[22m                                                                                                                                          |
|------|----------------------------------------------------------------------------------------------------------------------------------------------|
| bar1 | value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1 value1  |
|------|----------------------------------------------------------------------------------------------------------------------------------------------|
| bar2 | value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2 value2  |
+------+----------------------------------------------------------------------------------------------------------------------------------------------+

"
`;

exports[`DataPresenter collection of items tests collection of items should render using the supplied map() for a collection of items in a horizontal table 1`] = `
"+------+
| [1mfoo[22m  |
|------|
| bar1 |
|------|
| bar2 |
+------+

"
`;

exports[`DataPresenter single item should render a single item as json 1`] = `
"{
  \\"foo\\": \\"bar\\",
  \\"key\\": \\"value value value value value value value value value value value value value value value value value value value value \\"
}"
`;

exports[`DataPresenter single item should render a single item as vertical table using tableUserConfig 1`] = `
"+----------+------------------------------------------------------------------------------------------------------+
| [1mProperty[22m | [1mValue[22m                                                                                                |
|----------|------------------------------------------------------------------------------------------------------|
| foo      | \\"bar\\"                                                                                                |
|----------|------------------------------------------------------------------------------------------------------|
| key      | \\"value value value value value value value value value value value value value value value value val |
|          | ue value value value \\"                                                                               |
+----------+------------------------------------------------------------------------------------------------------+

"
`;

exports[`DataPresenter single item should render a single item vertical table 1`] = `
"+----------+----------------------------------------------------------------------------------------------------------------------------+
| [1mProperty[22m | [1mValue[22m                                                                                                                      |
|----------|----------------------------------------------------------------------------------------------------------------------------|
| foo      | \\"bar\\"                                                                                                                      |
|----------|----------------------------------------------------------------------------------------------------------------------------|
| key      | \\"value value value value value value value value value value value value value value value value value value value value \\" |
+----------+----------------------------------------------------------------------------------------------------------------------------+

"
`;
